File: blind75/Valid Anagram/Valid Anagram.java
class Solution {
    public boolean isAnagram(String s, String t) {
        if(s.length() != t.length()){
            return false;
        }

        HashMap<Character, Integer> counter = new HashMap<>();
        for(int i = 0; i < s.length(); i++){
            Character character = s.charAt(i);
            counter.put(character, counter.getOrDefault(character, 0) + 1);
        }

        for (int i = 0; i < t.length(); i++){
            Character character = t.charAt(i);
            if(counter.get(character) == null || counter.get(character) == 0){
                return false;
            }
            
            counter.put(character, counter.get(character) - 1);
        }

        return true;
    }
}

File: blind75/Valid Anagram/Valid Anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # strings must be equal length
        if len(s) != len(t):
            return False

        ## Brute Force
        # if sorted(s) != sorted(t): #O(nlogn + mlogm)
        #     return False

        # return True

        ## Optimal Solution
        counter = dict()

        for char in s:
            counter[char] = counter.get(char, 0) + 1

        for char in t:
            if char not in counter or counter[char] == 0:
                return False
            else:
                counter[char] = counter.get(char) - 1

        return True
        